Mathlib/LinearAlgebra/Basis/Exact.lean

```diff
@@ -0,0 +1,92 @@
+  apply (LinearIndependent.comp hli a hainj).map
+  rw [Submodule.disjoint_def, hfg.linearMap_ker_eq]
+  rintro - hy ⟨y, rfl⟩
+  have hz : s (f y) = 0 := by
+    revert hy
+    generalize f y = x
+    intro hy
+    induction' hy using Submodule.span_induction with m hm
+    · obtain ⟨i, rfl⟩ := hm
+      apply hsa
+    all_goals simp_all
+  replace hs := DFunLike.congr_fun hs y
+  simp only [LinearMap.coe_comp, Function.comp_apply, LinearMap.id_coe, id_eq] at hs
+  rw [← hs, hz, map_zero]
+
+private lemma top_le_span_of_aux (v : κ ⊕ σ → M)
+    (hg : Function.Surjective g) (hslzero : ∀ i, s (v (.inl i)) = 0)
+    (hli : LinearIndependent R (s ∘ v ∘ .inr)) (hsp : ⊤ ≤ Submodule.span R (Set.range v)) :
+    ⊤ ≤ Submodule.span R (Set.range <| g ∘ v ∘ .inl) := by
+  rintro p -
+  obtain ⟨m, rfl⟩ := hg p
+  wlog h : m ∈ LinearMap.ker s
+  · let x : M := f (s m)
+    rw [show g m = g (m - f (s m)) by simp [hfg.apply_apply_eq_zero]]
+    apply this hs hfg v hg hslzero hli hsp
+    replace hs := DFunLike.congr_fun hs (s m)
+    simp only [LinearMap.coe_comp, Function.comp_apply, LinearMap.id_coe, id_eq] at hs
+    simp [hs]
+  have : m ∈ Submodule.span R (Set.range v) := hsp trivial
+  obtain ⟨c, rfl⟩ := Finsupp.mem_span_range_iff_exists_finsupp.mp this
+  simp only [LinearMap.mem_ker, Finsupp.sum, map_sum, map_smul,
+    Finset.sum_sum_eq_sum_toLeft_add_sum_toRight, map_add, hslzero, smul_zero,
+    Finset.sum_const_zero, zero_add] at h
+  replace hli := (linearIndependent_iff'.mp hli) c.support.toRight (c ∘ .inr) h
+  simp only [Finset.mem_toRight, Finsupp.mem_support_iff, Function.comp_apply, not_imp_self] at hli
+  simp only [Finsupp.sum, Finset.sum_sum_eq_sum_toLeft_add_sum_toRight, hli, zero_smul,
+    Finset.sum_const_zero, add_zero, map_sum, map_smul]
+  exact Submodule.sum_mem _ (fun i hi ↦ Submodule.smul_mem _ _ <| Submodule.subset_span ⟨i, rfl⟩)
+
+lemma Submodule.top_le_span_of_exact_of_retraction (hg : Function.Surjective g)
+    (hsa : ∀ i, s (v (a i)) = 0) (hlib : LinearIndependent R (s ∘ v ∘ b))
+    (hab : Set.range (v ∘ a) ∪ Set.range (v ∘ b) = Set.range v)
+    (hsp : ⊤ ≤ Submodule.span R (Set.range v)) :
+    ⊤ ≤ Submodule.span R (Set.range <| g ∘ v ∘ a) := by
+  apply top_le_span_of_aux hs hfg (Sum.elim (v ∘ a) (v ∘ b)) hg hsa hlib
+  rwa [Set.Sum.elim_range, hab]
+
+/-- Let `0 → K → M → P → 0` be an exact sequence of `R`-modules, let `s : M → K` be a
+retraction of `f` and `v` be a basis of `M` indexed by `κ ⊕ σ`. Then
+if `s vᵢ = 0` for `i : κ` and `(s vⱼ)ⱼ` is linear independent for `j : σ`, then
+the images of `vᵢ` for `i : κ` form a basis of `P`.
+
+For convenience this is stated not for an arbitrary type `ι` with two maps `κ → ι` and `σ → ι`. -/
+noncomputable def Basis.ofSplitExact (hg : Function.Surjective g) (v : Basis ι R M)
+    (hainj : Function.Injective a) (hsa : ∀ i, s (v (a i)) = 0)
+    (hlib : LinearIndependent R (s ∘ v ∘ b))
+    (hab : Set.range (v ∘ a) ∪ Set.range (v ∘ b) = Set.range v) :
```

Comment:
I feel like `Codisjoint (Set.range a) (Set.range b)` will be easier to prove rather than `hab`.